[Upstream _NamedOptimzer] Reland PR (89480)

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):

Reland https://github.com/pytorch/pytorch/pull/89480/
* #90294
* __->__ #90293

Differential Revision: [D41783015](https://our.internmc.facebook.com/intern/diff/D41783015)

Commit:
Reland PR (89480)

[ghstack-poisoned]

fduwjj commented:
@pytorch rebase

fduwjj commented:
@pytorchbot rebase

Commit:
Update on "Reland PR (89480)"

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):

Reland https://github.com/pytorch/pytorch/pull/89480/
* #90294
* __->__ #90293



[ghstack-poisoned]

awgu commented on test/distributed/optim/test_named_optimizer.py:
```diff
@@ -0,0 +1,245 @@
+            "Expects the optim to be initialized before load but found not initialized"
+        )
+        with self.assertRaisesRegex(ValueError, err_msg):
+            named_optim_2.load_state_dict(state_dict_to_load)
```
nit: Add newline (might be required by linter).

fduwjj replied:
Linter does not complain previously. But adding it now.

awgu reviewed (approved):
I am stamping since I think this is the same as before just without `__all__`.

Please make sure `python test/test_public_bindings.py` passes and that the lint passes before merging. Thanks!

Commit:
Update on "[Upstream _NamedOptimzer] Reland PR (89480)"

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):

Reland https://github.com/pytorch/pytorch/pull/89480/
* #90294
* __->__ #90293



[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge

Commit:
Update on "[Upstream _NamedOptimzer] Reland PR (89480)"

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):

Reland https://github.com/pytorch/pytorch/pull/89480/
* #90294
* __->__ #90293



[ghstack-poisoned]

fduwjj commented:
@fduwjj has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41783015).
